File: L4/App_4_2/App_4_2.f90
!
! PCLP1 | L4
! Calcularea CMMDC-ului si al CMMMC-ului a doua numere.
!
! File: App_4_2.f90
!

PROGRAM main
    IMPLICIT NONE

    INTEGER*8 :: a, b

    WRITE(*, '(A)', ADVANCE='NO') "a = "
    READ(*,*) a

    WRITE(*, '(A)', ADVANCE='NO') "b = "
    READ(*,*) b
    
    WRITE(*,*) "cmmdc: ", cmmdc(a, b)
    WRITE(*,*) "cmmmc: ", cmmmc(a, b)

    CONTAINS

    FUNCTION cmmdc(a, b) RESULT(retval)
        IMPLICIT NONE

        INTEGER*8, INTENT(IN) :: a, b
        INTEGER*8 :: retval, r, ca, cb

        ca = abs(a)
        cb = abs(b)

        DO WHILE(cb .NE. 0)
            r = MOD(a, b)

            ca = b
            cb = r
        END DO

        retval = ca
    END FUNCTION cmmdc

    FUNCTION cmmmc(a, b) RESULT(retval)
        IMPLICIT NONE

        INTEGER*8 :: a, b, retval

        a = abs(a)
        b = abs(b)

        retval = a * b / cmmdc(a, b)
    END FUNCTION cmmmc
END PROGRAM main

File: L4/App_4_2/App_4_2.cpp
/**
 * @brief PCLP1 | L4
 *      Calcularea CMMDC-ului si al CMMMC-ului a doua numere.
 *
 * @file App_4_2.cpp
 */

#include <iostream>

uint64_t cmmdc(int a, int b)
{
    a = std::abs(a);
    b = std::abs(b);

    while (b != 0)
    {
        int r = a % b;
        a = b;
        b = r;
    }

    return static_cast<uint64_t>(a);
}

uint64_t cmmmc(int a, int b)
{
    a = std::abs(a);
    b = std::abs(b);

    uint64_t a64 = static_cast<uint64_t>(a);
    uint64_t b64 = static_cast<uint64_t>(b);

    return a64 * b64 / cmmdc(a, b);
}

int main()
{
    int a, b;

    std::cout << "a = "; std::cin >> a;
    std::cout << "b = "; std::cin >> b;

    std::cout << "cmmdc(a,b) = " << cmmdc(a, b) << "\ncmmmc(a,b) = " << cmmmc(a,b);

    return 0;
}